### PR TITLE
Return state changes in read only calls

### LINKED ISSUE #3401: Add state changes to read only call responses
Actually(testnet18), `StateChanges` aren't returned in the response of `execute_read_only_bytecode` and `execute_read_only_call`

As api responses are in `massa-models` and `StageChanges` are in `massa-final-state`module which rely on `massa-models, adding it naively to `ExecuteReadOnlyResponse` creates a circular dependency. 

We should find a better way to refactor this part to include `StateChanges` into `ExecuteReadOnlyResponse`.

related to #2950

### PR DESCRIPTION
**Pull request by aoudiamoncef**
* [x] document all added functions
* [x] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [x] add logs allowing easy debugging in case the changes caused problems
* [x] if the API has changed, update the API specification

closes #3401 and related (part 2) of #3383

* [x] fix wrong doc on `is_final` introduced in #3383
* [x] create `massa-api-module`
* [x] add `state_changes` to `ExecuteReadOnlyResponse`
* [ ] Create OpenRPC `StateChanges` object

### CODE REVIEW

**Comment by aoudiamoncef** [comment]
@AurelienFT is this payload correct ? if no, how can I verify it ? 

```
                "pos_changes": {
                    "seed_bits": {
                        "order": "bitvec::order::Lsb0",
                        "head": {
                            "width": 8,
                            "index": 0
                        },
                        "bits": 0,
                        "data": []
                    },
                    "roll_changes": {},
                    "production_stats": {},
                    "deferred_credits": {
                        "credits": {},
                        "hash": "11111111111111111111111111111111273Yts"
                    }
                },
```

As I have a lot empty fields, it's very hard to create an OpenRPC object. I tried to do it from the Rust structs, it's harder as I have to know every serialiser.

**Comment by AurelienFT** [comment]
Regarding https://github.com/massalabs/massa/pull/3405#issuecomment-1380712297 maybe we should skip serializing some fields : 

- seed_bits
- executed_ops_changes
- deferred_credits

The problem would be that in the future if we want to serialize to json this structure in an other context they will be skipped also. I don't think it we should do it then.

No problems if it's not defined in OpenRPC it's a complex object if we host our Rust doc, one day, we will add a link to the doc of this structure

**Review by AurelienFT — COMMENTED** [review]
Nice refactor ! Just few things that could help readability of the project architecture

File: massa-api-exports/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+strum = { version = "0.24", features = ["derive"] }
+
+# custom modules
+massa_signature = { path = "../massa-signature" }
+massa_time = { path = "../massa-time" }
+massa_models = { path = "../massa-models" }
+massa_final_state = { path = "../massa-final-state" }
+
+[dev-dependencies]
+serial_test = "0.10"
```
[the remark this answers is not in the document]
Comment (aoudiamoncef) [reply massa-api-exports/Cargo.toml:18]: I hope, we will add some tests but for now, there is no valable reason.

File: massa-models/src/api.rs [hunk 2]
```diff
@@ -1,144 +1,7 @@
 // Copyright (c) 2022 MASSA LABS <info@massa.net>
```
[the remark this answers is not in the document]
Comment (aoudiamoncef) [reply massa-models/src/api.rs:1]: Yeah, it creates circular deps. It should be fine to split it.

File: massa-api/src/public.rs [hunk 3]
```diff
@@ -188,7 +187,9 @@ impl MassaRpcServer for API<Public> {
                 ),
                 gas_cost: result.as_ref().map_or_else(|_| 0, |v| v.gas_cost),
                 output_events: result
-                    .map_or_else(|_| Default::default(), |mut v| v.out.events.take()),
+                    .as_ref()
+                    .map_or_else(|_| Default::default(), |v| v.out.events.clone().take()),
```
[the remark this answers is not in the document]
Comment (aoudiamoncef) [reply massa-api/src/public.rs:191]: l give it a try, rustc decides

**Comment by aoudiamoncef** [comment]
> Regarding [#3405 (comment)](https://github.com/massalabs/massa/pull/3405#issuecomment-1380712297) maybe we should skip serializing some fields :
> 
> * seed_bits
> * executed_ops_changes
> * deferred_credits
> 
> The problem would be that in the future if we want to serialize to json this structure in an other context they will be skipped also. I don't think it we should do it then.
> 
> No problems if it's not defined in OpenRPC it's a complex object if we host our Rust doc, one day, we will add a link to the doc of this structure

I'm not a big fan of hardcoded behaviours in simples structs. May be will create a light objects to choose only what we need ?

File: massa-api-exports/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (aoudiamoncef) [reply massa-api-exports/Cargo.toml:18]: done

File: massa-api/src/public.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (aoudiamoncef) [reply massa-api/src/public.rs:191]:
```rust
error[E0308]: mismatched types
   --> massa-api/src/public.rs:189:32
    |
189 |                   output_events: result
    |  ________________________________^
190 | |                     .map_or_else(|_| Default::default(), |v| v.out.events.clone()),
    | |__________________________________________________________________________________^ expected struct `VecDeque`, found struct `EventStore`
    |
    = note: expected struct `VecDeque<SCOutputEvent>`
               found struct `EventStore`
```

[the remark this answers is not in the document]
Comment (AurelienFT) [reply massa-api/src/public.rs:191]: `v.out.events.clone().0` should work `EventStore` is a tuple structure and his inner element is public you should be able to use it directly.

**Comment by AurelienFT** [comment]
> I'm not a big fan of hardcoded behaviours in simples structs. May be will create a light objects to choose only what we need ?

Yes it could be better to have a light structure that we use to export only some part of the big one. Let's make it in another issue it's not a priority and we have a lot to do in this testnet on the API side.

File: massa-api/src/public.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (aoudiamoncef) [reply massa-api/src/public.rs:191]: done

File: massa-api-exports/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,613 @@
+// Copyright (c) 2022 MASSA LABS <info@massa.net>
```
[the remark this answers is not in the document]
Comment (aoudiamoncef) [reply massa-api-exports/src/lib.rs]: Done

**Comment by aoudiamoncef** [comment]
@AurelienFT,

IMHO, to be coherent with other modules, we should move traits from massa-api to massa-api-export.

May be we should create a submodule which wraps all DTOs to massa-api-export/dtos.

**Comment by AurelienFT** [comment]
> @AurelienFT,
> 
> IMHO, to be coherent with other modules, we should move traits from massa-api to massa-api-export.
> 
> May be we should create a submodule which wraps all DTOs to massa-api-export/dtos.

We only move traits when they are created to be exposed to other module which is not the case in the API it's just for Public and Private

**Review by AurelienFT — APPROVED** [review]
This is way more clean like that. The massa-models module can now breath.

**Comment by aoudiamoncef** [comment]
Thanks for the clarification. let's get a last review before merging

**Pull request merged**